### ISSUE TITLE
transformers backend: audio encoder attention window (cu_seqlens) is not applied on sdpa/eager; clips > 8 s are encoded with full attention

### ISSUE THREAD

**Issue by moona3k** [QwenLM/Qwen3-ASR#213]
## Summary

In the `transformers_backend`, `Qwen3ASRAudioEncoder.forward` computes `cu_seqlens` for the `n_window_infer` attention windows but never applies them unless the attention implementation is `flash_attention_2`. The method that builds the block-diagonal mask, `_prepare_attention_mask`, is defined and never called. With `sdpa` (the default on CPU and on GPU when FA2 is not installed) or `eager`, every audio token attends to every other token in the clip, so any input longer than one window (`n_window_infer = 800` mel frames = 8 s = 104 encoder tokens) is encoded differently from how the model was trained and from what the FA2 and vLLM paths compute.

## Where

At `7c6daf7` (current `main`, package version 0.0.6), `qwen_asr/core/transformers_backend/modeling_qwen3_asr.py`:

- `_prepare_attention_mask` is defined at [L649](https://github.com/QwenLM/Qwen3-ASR/blob/7c6daf7/qwen_asr/core/transformers_backend/modeling_qwen3_asr.py#L649) and returns `None` only for `flash_attention_2`; it is the only occurrence of that name in the file.
- `forward` calls each layer as `encoder_layer(hidden_states, cu_seqlens)` at [L729](https://github.com/QwenLM/Qwen3-ASR/blob/7c6daf7/qwen_asr/core/transformers_backend/modeling_qwen3_asr.py#L729) with `attention_mask` left at `None`.
- `Qwen3ASRAudioAttention.forward` ([L477](https://github.com/QwenLM/Qwen3-ASR/blob/7c6daf7/qwen_asr/core/transformers_backend/modeling_qwen3_asr.py#L477)) passes `cu_seqlens` only as `cu_seq_lens_q/k` (consumed by the FA2 kernel) and `attention_mask=None` to the sdpa/eager interface, which then attends over the full sequence.

For comparison, the native implementation on `transformers` `main` (`src/transformers/models/qwen3_asr/modeling_qwen3_asr.py`, `Qwen3ASRAudioAttention.forward`) has an explicit non-flash branch that splits Q/K/V at `cu_seqlens` and runs attention per window ("Other implementations: Process each chunk separately"). The vLLM backend in this repo also windows via `cu_seqlens`. Only the bundled transformers backend on sdpa/eager does not.

## Reproduction

Any clip longer than 8 s; below uses LibriSpeech `test-clean/1089/134691/1089-134691-0011.flac` (20.0 s). The patch is two lines: pass the mask the encoder already knows how to build.

```python
import sys, numpy as np, soundfile as sf, torch
from qwen_asr import Qwen3ASRModel

audio, sr = sf.read(sys.argv[1], dtype="float32"); assert sr == 16000
m = Qwen3ASRModel.from_pretrained("Qwen/Qwen3-ASR-0.6B", device_map="cpu", dtype=torch.float32)
enc = m.model.thinker.audio_tower
print("attn_implementation:", enc.config._attn_implementation)
feats = m.processor.feature_extractor(audio, sampling_rate=16000, return_tensors="pt",
                                      padding="do_not_pad", return_attention_mask=True)
encode = lambda: m.model.thinker.get_audio_features(feats["input_features"], feats["attention_mask"])

with torch.inference_mode():
    as_shipped = encode()

for layer in enc.layers:  # apply the mask that forward() never passes
    orig = layer.forward
    def forward(h, cu_seqlens, attention_mask=None, _orig=orig, **kw):
        if attention_mask is None:
            attention_mask = enc._prepare_attention_mask(h, cu_seqlens)
        return _orig(h, cu_seqlens, attention_mask=attention_mask, **kw)
    layer.forward = forward

with torch.inference_mode():
    windowed = encode()

d = (as_shipped - windowed).abs()
print(f"tokens={as_shipped.shape[0]}  mean|diff|={d.mean():.4f}  max={d.max():.4f}  "
      f"relative={d.mean()/windowed.abs().mean():.1%}")
```

Output (macOS 26.6, Apple M4 Pro, CPU, fp32; identical on torch 2.10.0 and 2.14.0, transformers 4.57.6):

```
attn_implementation: sdpa
tokens=261  mean|diff|=0.0036  max=0.1380  relative=24.9%
```

The per-window mean difference is flat across the three windows (0.0036, 0.0033, 0.0041), so this is not an edge effect at the boundary; the whole encoding changes.

Below one window the two paths are identical: on 33 LibriSpeech clips <= 8 s the encoder outputs match to fp32 rounding, on every one of 17 clips > 8 s they diverge, and the divergence grows with duration.

## Effect on transcripts

Running `Qwen3ASRModel.transcribe` both ways on the 22 clips > 8 s among the first 60 of LibriSpeech test-clean: 4 transcripts change. In this English sample the changes are punctuation and casing (`dinner, turnips` -> `dinner: turnips`; `little office` -> `Little Office`; `certitude that` -> `certitude, that`; `stone, poised` -> `stone poised`). The words happen to survive here; I have not measured the effect on other languages or noisier audio.

Word-level timestamps from `Qwen3-ForcedAligner-0.6B` are also computed through this encoder; on 50 clips the timestamps agree to ~5.6 ms either way because the 80 ms argmax absorbs the encoder change.

## Why it matters

1. The same package gives different encoder outputs for the same audio depending on whether FA2 is installed. sdpa/eager users (CPU, MPS, GPUs without FA2) get a model that was not trained this way for anything over 8 s.
2. Anyone using the transformers backend as a numerical reference (parity tests, ports, quantization checks) will see a spurious 5-25% encoder divergence on long clips. This is how I found it: an MLX port that applies the window matched the reference exactly on clips under 8 s and diverged on every clip over 8 s until the mask was patched in.

## Suggested fix

Either of:

- In `Qwen3ASRAudioEncoder.forward`, build the mask once and pass it: `attention_mask = self._prepare_attention_mask(hidden_states, cu_seqlens)` before the layer loop, then `encoder_layer(hidden_states, cu_seqlens, attention_mask=attention_mask)`. This uses the existing method and is what the two-line patch above does. It materializes an `(L, L)` mask, which is fine for the 30 s inputs the processor produces (390 tokens).
- Or mirror `transformers` `main`: in `Qwen3ASRAudioAttention.forward`, when not on flash attention, split Q/K/V by `cu_seqlens` and run attention per window, then concatenate. No dense mask, same result.

Happy to open a PR for either if you have a preference.

**Comment by moona3k** [comment]
Ah, there already exists an [open PR](https://github.com/QwenLM/Qwen3-ASR/pull/103) for the fix, though it's been sitting since Feb 2026.

```
// local agent verification
- On a 20 s clip (CPU, fp32, `sdpa`, torch 2.10.0): as-shipped vs dense block-diagonal mask, mean|diff| = 0.0025, max = 0.0955, relative = 18.5%.
- Dense block-diagonal mask vs an independent per-window split (Q/K/V sliced at `cu_seqlens`, SDPA per window, concatenated, i.e. the `transformers` `main` recipe): mean|diff| = 0.00000, max = 0.0000.
```

**Comment by Mark-Oliass** [comment]
Confirming this from a production deployment, with a much more severe symptom than the parity
divergence reported above.

**Setup.** We run the bundled `transformers_backend` (package 0.0.6, `_attn_implementation = "sdpa"`,
`flash_attn` not installed) on long meeting recordings, sliced into ~180 s chunks. The audio encoder
is called once per chunk, so N ≈ 2331 encoder positions per call.

**Symptom.** The decoder does not merely change punctuation — it enters a degenerate repetition
loop: the same 2–11 character unit is repeated hundreds of times until it hits `max_new_tokens`
(a 3-character unit, or even a single character — the classic attractor, not a boundary effect).
Downstream, our repetition filter then collapses that text, so most of the transcript for the chunk
is silently destroyed.

**Scale.** 6 meetings / 12 audio variants / 48–49 speech blocks. Criterion: the longest
periodically-repeating span covers > 30 % of the chunk's decoded text.

| | runaway blocks / speech blocks |
| --- | --- |
| as shipped (`sdpa`) | 18/48 = **37.5 %** |
| with the mask passed (PR #103) | 7/49 = **14.3 %** |
| vLLM backend, same audio (already windows via `cu_seqlens`) | 5/49 = 10.2 % |

After excluding audio that a human listener also cannot transcribe (near-silence / unintelligible
recordings), the fixed path and the vLLM path are both **0/39**.

**Independent evidence that the window is not applied.** Instead of comparing full encoder outputs,
we perturb the mel frames of the *last* window and measure the change in the *first* window's
output. As shipped the first window changes (max abs diff 0.1436); with the mask passed it is
exactly 0.0. The leakage is real, and the mask removes it — same conclusion as your `max=0.1380`
on a 20 s clip.

**Your two proposed fixes — a data point for the choice.** We measured the dense-mask variant:

- mask cost ≈ 11 MB for N ≈ 2331 (a 180 s chunk), built once per encoder call and reused across
  all 24 layers;
- encoder attention time goes 22384 µs → 29436 µs (+31 %), because passing a mask moves SDPA off
  the flash backend onto the memory-efficient one.

Both are acceptable at these lengths, but the no-dense-mask route (split Q/K/V at `cu_seqlens` and
run per window, as `transformers` main does) would avoid both.

**Cross-check on the mask variant.** We also installed flash-attn and ran
`attn_implementation="flash_attention_2"` on the same audio. Merged text, merged-text SHA256,
per-chunk text, per-chunk token counts and per-chunk raw-text SHA256 are **all identical** (5/5
chunks). So on this input the block-diagonal mask is not merely "approximate" relative to FA2 — it
is exactly equivalent, consistent with the note in `_prepare_attention_mask`.

**On the reported impact.** You noted the words survive on the English LibriSpeech sample and that
other languages / noisier audio were not measured. Our domain is Chinese industrial meeting
recordings with heavy noise, and there the failure mode is the full repetition loop above. So the
parity divergence you measured is real, but the production impact can be far larger than those
punctuation changes suggest.

PR #103 implements the mask variant and matches what we run in production.
